rainier-cats/src/main/scala/com/stripe/rainier/cats/package.scala

```diff
@@ -28,17 +27,21 @@ private[cats] class MonadGenerator extends Monad[Generator] {
   def flatMap[A, B](fa: Generator[A])(f: A => Generator[B]): Generator[B] =
     fa.flatMap(f)
 
-  // note: currently not stack safe
-  def tailRecM[A, B](a: A)(f: A => Generator[Either[A, B]]): Generator[B] =
-    new Generator[B] {
-      lazy val step = f(a)
-      def requirements: Set[Real] = step.requirements
-      def get(implicit r: RNG, n: Numeric[Real]): B =
-        step.get match {
-          case Right(b) => b
-          case Left(aa) => tailRecM(aa)(f).get
-        }
-    }
+  def tailRecM[A, B](a: A)(f: A => Generator[Either[A, B]]): Generator[B] = {
```

Comment:
I think this is a faithful translation of the original tailRecM, but I don't think the original tailRecM is correct.

Note, after the first application of `f` we are discarding the requirements. I think we need to be accumulating them in the loop.

Something like:
```scala
def run(a: A, reqs: Set[Real], r: RNG, n: Numeric[Real]): (Set[Real], B) =
  f(a) match {
    case Generator.Const(r1, Right(b)) => (reqs ++ r1, b)
    case Generator.Const(r1, Left(a)) => run(a, reqs ++ r1, r, n)
    case Generator.From(r1, fromFn) =>
      fromFn(r, n) match {
        case Right(b) => (reqs ++ r1, b)
        case Left(a) => run(a, reqs ++ r1, r, n)
      }
  }
```
But that leaves open what to do with all these accumulated requirements?

That said, looking at the original flatMap, it also discards the requirements for all but the first in the chain. That seems really strange to me....

Reply:
yeah, it's super strange to me too. I added a note to the `flatMap` implementation about that (you probably saw it). I didn't see a clear way to get the requirements up and out of the body of `From`'s function; that's the challenge that was maybe ignored in the first place, that pops up in `flatMap` of course. I just don't know enough about the reason for `requirements`.

Reply:
Funny, btw, I had that exact signature and implementation but cut it out with no mention...

Reply:
Wait, I can fix it. Hold tight. (Will delete this comment after.)

Reply:
I'll dig into this PR in more detail later, but to address the deal with `requirements`: the idea is that to the extent that we can anticipate `Real` objects that will be converted to double using the `Numeric[Real]` passed to `get`, we can pre-compute them more efficiently using the Real->JVM bytecode compiler; otherwise we have to fallback dynamically to a Real interpreter. So accumulating `requirements` is a very useful optimization, but semantically it's fine to discard in any case where we can't figure out how to do it.

Reply:
We could keep them if we made them a return value of `def get(r: RNG, n: Numeric[Real]): (Set[Real], A)` would that defeat the purpose because you have to call get first?

Reply:
It would yep.

Reply:
This will do it for `tailRecM`... looking again for a way to do it without calling `get` in the `flatMap` implementation.

```
  def tailRecM[A, B](a: A)(f: A => Generator[Either[A, B]]): Generator[B] = {
    @tailrec
    def run(g: Generator[Either[A, B]],
            reqs: Set[Real],
            r: RNG,
            n: Numeric[Real]): (Set[Real], B) =
      g match {
        case Generator.Const(r1, Left(a))  => run(f(a), reqs ++ r1, r, n)
        case Generator.Const(r1, Right(b)) => (reqs ++ r1, b)
        case Generator.From(r1, fromFn) =>
          fromFn(r, n) match {
            case Left(a)  => run(f(a), reqs ++ r1, r, n)
            case Right(b) => (reqs ++ r1, b)
          }
      }
    val step = f(a)
    Generator
      .require(step.requirements)(run(step, Set.empty, _, _))
      .flatMap { case (reqs, u) => Generator.Const(reqs, u) }
  }
```

Reply:
I don't know/grok `tailRecM` well enough to speak to it, but in general it'll be impossible to accumulate `requirements` across a `flatMap` in any way that's effective; this is the standard thing of not being able to statically analyze a monad (vs an applicative).

Reply:
actually, @avi-stripe, of course that is right - this just kicks it back to the `flatMap` implementation, which in my comment above I note doesn't work yet. (Because it can't work.) I'll forget it and leave the code as is.